File: code/smv-debugger/src/main/java/org/fbme/smvDebugger/panel/mvc/DebugPanelModel.java
package org.fbme.smvDebugger.panel.mvc;

import javax.swing.SpinnerModel;
import javax.swing.table.TableModel;
import javax.swing.ListSelectionModel;
import javax.swing.BoundedRangeModel;
import org.fbme.smvDebugger.model.Counterexample;
import javax.swing.SpinnerListModel;
import org.fbme.smvDebugger.commons.ArrayUtils;
import javax.swing.DefaultListSelectionModel;
import javax.swing.DefaultBoundedRangeModel;

public class DebugPanelModel {
  private static final String FIRST_COLUMN_NAME = "Components/States";

  private final StepIndexModel stepIndexModel;
  private final SpinnerModel stepSpinnerModel;
  private final GlobalTimeModel globalTimeModel;
  private final ConditionModel conditionModel;
  private final TableModel itemModel;
  private final TableModel valueModel;
  private final ListSelectionModel dataSelectionModel;
  private final BoundedRangeModel dataScrollModel;

  public DebugPanelModel(final Counterexample counterexample) {
    stepIndexModel = new StepIndexModel();

    stepSpinnerModel = new SpinnerListModel(counterexample.getSteps());
    globalTimeModel = new GlobalTimeModel();
    conditionModel = new ConditionModel();

    itemModel = new UneditableTableModel(ArrayUtils.to2dArray(counterexample.getItems()), new String[]{FIRST_COLUMN_NAME});
    valueModel = new UneditableTableModel(counterexample.getValues(), counterexample.getSteps());

    dataSelectionModel = new DefaultListSelectionModel();
    dataSelectionModel.setSelectionMode(ListSelectionModel.SINGLE_SELECTION);

    dataScrollModel = new DefaultBoundedRangeModel();
  }

  public StepIndexModel getStepIndexModel() {
    return stepIndexModel;
  }

  public SpinnerModel getStepSpinnerModel() {
    return stepSpinnerModel;
  }

  public GlobalTimeModel getGlobalTimeModel() {
    return globalTimeModel;
  }

  public ConditionModel getConditionModel() {
    return conditionModel;
  }

  public TableModel getItemModel() {
    return itemModel;
  }

  public TableModel getValueModel() {
    return valueModel;
  }

  public ListSelectionModel getDataSelectionModel() {
    return dataSelectionModel;
  }

  public BoundedRangeModel getDataScrollModel() {
    return dataScrollModel;
  }
}

File: code/smv-debugger/src/main/kotlin/org/fbme/smvDebugger/panel/mvc/DebugPanelModel.kt
package org.fbme.smvDebugger.panel.mvc

import org.fbme.smvDebugger.model.Counterexample
import org.fbme.smvDebugger.model.SystemItem
import javax.swing.*
import javax.swing.table.TableModel

class DebugPanelModel(counterexample: Counterexample?) {
    val stepIndexModel: StepIndexModel = StepIndexModel()
    val stepSpinnerModel: SpinnerModel
    val globalTimeModel: GlobalTimeModel
    val conditionModel: ConditionModel
    val itemModel: TableModel
    val valueModel: TableModel
    val dataSelectionModel: ListSelectionModel
    val dataScrollModel: BoundedRangeModel

    companion object {
        private const val FIRST_COLUMN_NAME = "Components/States"
    }

    init {
        stepSpinnerModel = SpinnerListModel(counterexample!!.steps)
        globalTimeModel = GlobalTimeModel()
        conditionModel = ConditionModel()
        val elements: Array<SystemItem> = counterexample.items
        val arr = emptyArray<Array<SystemItem>>().plus(elements)
        itemModel = UneditableTableModel(arr, arrayOf<String?>(FIRST_COLUMN_NAME))
        valueModel = UneditableTableModel(counterexample.values, counterexample.steps)
        dataSelectionModel = DefaultListSelectionModel()
        dataSelectionModel.selectionMode = ListSelectionModel.SINGLE_SELECTION
        dataScrollModel = DefaultBoundedRangeModel()
    }
}
